haystack/nodes/ranker/cohere.py

```diff
@@ -0,0 +1,227 @@
+import json
+import os
+from typing import Optional, Dict, Union, List, Any
+import logging
+
+import requests
+
+from haystack.environment import HAYSTACK_REMOTE_API_TIMEOUT_SEC, HAYSTACK_REMOTE_API_MAX_RETRIES
+from haystack.errors import CohereInferenceLimitError, CohereUnauthorizedError, CohereError
+from haystack.utils import request_with_retry
+
+from haystack.errors import HaystackError
+from haystack.schema import Document
+from haystack.nodes.ranker.base import BaseRanker
+
+
+logger = logging.getLogger(__name__)
+
+
+TIMEOUT = float(os.environ.get(HAYSTACK_REMOTE_API_TIMEOUT_SEC, 30))
+RETRIES = int(os.environ.get(HAYSTACK_REMOTE_API_MAX_RETRIES, 5))
+
+
+class CohereRanker(BaseRanker):
+    """
+    Re-Ranking can be used on top of a retriever to boost the performance for document search.
+    This is particularly useful if the retriever has a high recall but is bad in sorting the documents by relevance.
+
+    Cohere models are trained with a context length of 510 tokens - the model takes into account both the input
+    from the query and document. If your query is larger than 256 tokens, it will be truncated to the first 256 tokens.
+
+    Cohere breaks documents into 510 token chunks. For example, if your query is 50 tokens and your document is
+    1024 tokens, your document will be broken into the following chunks:
+    ```bash
+    relevance_score_1 = <query[0,50], document[0,460]>
+    relevance_score_2 = <query[0,50], document[460,920]>
+    relevance_score_3 = <query[0,50],document[920,1024]>
+    relevance_score = max(relevance_score_1, relevance_score_2, relevance_score_3)
+    ```
+
+    Find more Cohere reranking best practices at https://docs.cohere.com/docs/reranking-best-practices.
+    """
+
+    def __init__(
+        self,
+        api_key: str,
+        model_name_or_path: str,
+        top_k: int = 10,
+        return_documents: bool = False,
+        max_chunks_per_doc: Optional[int] = None,
+    ):
+        """
+         Creates an instance of CohereInvocationLayer for the specified Cohere model.
+
+        :param api_key: Cohere API key
+        :param model_name_or_path: Cohere model name. See supported models at https://docs.cohere.com/docs/models.
+        :param top_k: The maximum number of documents to return.
+        :param return_documents: If false, returns results without the doc text - the api will return a list of
+            {index, relevance score} where index is inferred from the list passed into the request.
+            If true, returns results with the doc text passed in - the api will return an ordered list of
+            {index, text, relevance score} where index + text refers to the list passed into the request.
+        :param max_chunks_per_doc: If your document exceeds 512 tokens, this will determine the maximum number of
+            chunks a document can be split into. For example, if your document is 6000 tokens, with the default of 10,
```

Comment:
Maybe we can make it more explicit that 10 is the default of the Cohere API for max_chunks_per_doc.